services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/attachment/application/ApplicationAttachmentContentTransferService.kt

```diff
@@ -0,0 +1,55 @@
+package fi.hel.haitaton.hanke.attachment.application
+
+import fi.hel.haitaton.hanke.attachment.application.ApplicationAttachmentContentService.Companion.generateBlobPath
+import fi.hel.haitaton.hanke.attachment.azure.Container
+import fi.hel.haitaton.hanke.attachment.common.ApplicationAttachmentContentRepository
+import fi.hel.haitaton.hanke.attachment.common.ApplicationAttachmentEntity
+import fi.hel.haitaton.hanke.attachment.common.ApplicationAttachmentRepository
+import fi.hel.haitaton.hanke.attachment.common.AttachmentContentNotFoundException
+import fi.hel.haitaton.hanke.attachment.common.FileClient
+import mu.KotlinLogging
+import org.springframework.data.repository.findByIdOrNull
+import org.springframework.http.MediaType
+import org.springframework.stereotype.Service
+import org.springframework.transaction.annotation.Transactional
+
+private val logger = KotlinLogging.logger {}
+
+@Service
+class ApplicationAttachmentContentTransferService(
+    private val attachmentRepository: ApplicationAttachmentRepository,
+    private val attachmentContentRepository: ApplicationAttachmentContentRepository,
+    private val fileClient: FileClient,
+) {
+    fun nextTransferableAttachment(): ApplicationAttachmentEntity? =
+        attachmentRepository.findByBlobLocationIsNull().firstOrNull()
```

Comment:
Passing entities beyond transaction boundaries can lead to difficult bugs. I think this might be safe, but I think the code would work just as well if this method returned the domain object.
```suggestion
    fun nextTransferableAttachment(): ApplicationAttachmentMetadata? =
        attachmentRepository.findFirstByBlobLocationIsNull()?.toDomain()
```

Reply:
The difference would be that in the end in `cleanUpDatabase`, you would need to load the entity from the database, instead of an explicit save:
```kotlin
        val attachmentEntity = attachmentRepository.getReferenceById(attachment.id)
        logger.info { "Updating attachment ${attachmentEntity.id} with blob location $blobPath" }
        attachmentEntity.blobLocation = blobPath
```

Reply:
You are of course right. I haven't been working with JPA and Entities for while and it shows in things like this. After refactoring to follow Hanke attachment migration I hope this kind of errors are no more.